woocommerce/woocommerce-gutenberg-products-block
# Ensure cart changes remain after using back button to get back to the cart

### Description
If the customer alters their cart, i.e. changes the quantity or adds an item, then navigates away from the page, for example by clicking the "Proceed to checkout" button, and then returns to the cart via the back button in their browser, then their changes will not be shown. The client and the server are out of sync.

This happens because we print the cart data onto the page at page load, and the cart hydrates itself based on this. When the user returns to the page using the back button, the data that were printed onto the page are stale, so the cart hydrates itself with the data that were present when the page was first loaded.

### Changes made

#### Source Code
- Added a new action to `assets/js/data/cart/actions` called: `cartDataIsStale`. This will set a flag in the store indicating the cart data is out of sync with the server. This is based on the check detailed below.
- Add `cartUpdateMiddleware` to `assets/js/middleware/cart-update.ts` which will write the timestamp of the last cart update to the browser's local storage.
- Update the `CartSchema` to add a new field: `generated_timestamp` this will allow us to know when the data were generated so we can compare it to the last time the cart was updated (which is stored in local storage).
- Add a check to the `withStoreCartApiHydration` hook that will compare the `generated_timestamp` time to the one in local storage to determine if the cart needs to be fetched from the API. 
- The cart data will not be rehydrated from `wcSettings` if it is determined that the data is stale.

#### Test code
This PR also introduces some new e2e tests to test this functionality. I needed to create some utility functions to enable testing the front end. The details of this follow.

- Install `@woocommerce/e2e-utils` to the project as a dev dependency. This is so we can use the `shopper.addToCart` flow.
- Add `visitPostOfType` function. This will cause the e2e browser to visit a page in the WordPress dashboard by name and type. This is needed so we can navigate to a `product` page by name. It could also be useful for other tests in future.
- Add `getBlockPagePermalink` function - this will cause the e2e browser to return the permalink of the page it is on if it the page is using the block editor. The analogue to this function is `getNormalPagePermalink` which will do the same but on a page that doesn't use the block editor (for example a product page in WooCommerce).
- Add tests for the cart on the front-end.


Fixes #3827

### How to test the changes in this Pull Request:

1. Add items to your cart then go to the cart block.
2. Change the quantity of an item, you may try changing multiple items, with various different quantities. **Make a note of the values!**
3. Press the "Proceed to checkout" button.
4. Use the browser's back button to get back to your cart.
5. Ensure the quantities displayed in the cart match what you expect them to from step 2.
6. Run the e2e tests and ensure all suites still pass.



### Changelog

> Ensure the cart stays up to date if the user navigates using the back/forward buttons in the browser.
